index.js

```diff
@@ -2,6 +2,112 @@
 var strictUriEncode = require('strict-uri-encode');
 var objectAssign = require('object-assign');
 
+function encodeArrayFormat(opts) {
+	switch (opts.arrayFormat) {
+		case 'index':
+			return function (key, value, index) {
+				return value === null ? [
+					encode(key, opts),
+					'[',
+					index,
+					']'
+				].join('') : [
+					encode(key, opts),
+					'[',
+					encode(index, opts),
+					']=',
+					encode(value, opts)
+				].join('');
+			};
+
+		case 'bracket':
+			return function (key, value) {
+				return value === null ? encode(key, opts) : [
+					encode(key, opts),
+					'[]=',
+					encode(value, opts)
+				].join('');
+			};
+
+		default:
+			return function (key, value) {
+				return value === null ? encode(key, opts) : [
+					encode(key, opts),
+					'=',
+					encode(value, opts)
+				].join('');
+			};
+	}
+}
+
+function decodeArrayFormat(opts) {
+	var index;
+	var list;
+
+	switch (opts.arrayFormat) {
+		case 'index':
+			return function (key, value, accumulator) {
+				if (arguments.length === 1) {
```

Comment:
Maybe we should simplify this a bit:

`decodeArrayFormat` -> `parserForArrayFormat` which returns a parser function that has the following signature`function (key, value, params) {}`.

Then, in `parse(), this function is simply called once with `key`, `value` and `ret`.